stdnum/gs1_128.py

```diff
@@ -0,0 +1,304 @@
+    '8102': (str, 2, True),
+    '8110': (str, 30, True),
+    '8200': (str, 70, True),
+    '90': (str, 30, True),
+    '91': (str, 90, True),
+    '92': (str, 90, True),
+    '93': (str, 90, True),
+    '94': (str, 90, True),
+    '95': (str, 90, True),
+    '96': (str, 90, True),
+    '97': (str, 90, True),
+    '98': (str, 90, True),
+    '99': (str, 90, True),
+}
+
+
+def _convert_date(val):
+    return datetime.strptime(val, '%y%m%d').date()
+
+
+def _convert_int(val):
+    return int(val)
+
+
+def _convert_float(val):
+    digits = float(_convert_int(val[:2]))
+    value = float(_convert_int(val[2:]))
+    return value / 10 ** digits
+
+
+CONVERTERS = {
+    date: _convert_date,
+    int: _convert_int,
+    float: _convert_float,
+}
+
+
+def _encode_string(val, length):
+    return val.ljust(length)
+
+
+def _encode_date(val, length):
+    return val.strftime('%y%m%d')
+
+
+def _encode_int(val, length):
+    return _encode_string(str(val), length)
+
+
+def _encode_float(val, length):
+    number, decimals = str(val).split('.')
+    digits = str(len(decimals)).rjust(2, '0')
+    number = number.rjust(6 - len(decimals), '0')
+    return _encode_string(digits + number + decimals, length)
+
+
+ENCODERS = {
+    str: _encode_string,
+    date: _encode_date,
+    int: _encode_int,
+    float: _encode_float,
+}
+
+
+def compact(number):
+    """
+    Convert the GS1-128 to the minimal representation.
+
+    This strips the number of any valid separators and removes surrounding
+    whitespace.
+    """
+    return clean(number, '()').strip()
+
+
+def info(code):
+    """
+    Return a dictionary containing the gs1-128 info embeded in code
+
+    Each application code's data is converted to the proper python:
+        * str: for string values
+        * datetime.date: for date values
+        * int: for int values
+        * float: for numeric values
+
+    Escape characters are also removed
+    """
+    code = compact(code)
+    queue = deque(code)
+
+    data = {}
+    identifier = ''
+    while queue:
+        identifier += queue.popleft()
+        if identifier in APPLICATION_CODES:
+            converter, length, variable = APPLICATION_CODES[identifier]
+            value = ''
+            for _ in range(length):
+                try:
+                    char = queue.popleft()
+                    # If it's a stop char do not continue
+                    if variable and char == chr(206):
+                        break
+                    value += char
+                except IndexError:
+                    pass
+            data[identifier] = CONVERTERS.get(converter, lambda a: a)(value)
+            identifier = ''
+    return data
+
+
+def encode(data):
+    """
+    Return a GS1-128 code for application identifiers contained in the data
+    dictionary
+
+    Use the compacted version of the code
+    """
+    code = ''
+    # Use sorted to keep a logical order of keys
+    for key in sorted(data.keys()):
```

Comment:
As the length of the key names is variable, I think it will be better to sort them with left padding filled with 0.